bentoml/_internal/runner/runner.py

```diff
@@ -152,21 +202,70 @@ def batch_options(self) -> BatchOptions:
     def _impl(self) -> "RunnerImpl":
         return create_runner_impl(self)
 
+    @final
+    def _execute_pre_run_hooks(self, *args, **kwargs) -> None:
+        """
+        Execute all pre-run hooks.
+        """
+
+        for hook in self._pre_run_hooks.values():
+            hook(*args, **kwargs)
+
+    @final
+    def _execute_post_run_hooks(self, *args, **kwargs) -> None:
+        """
+        Execute all post-run hooks.
+        """
+
+        for hook in self._post_run_hooks.values():
+            hook(*args, **kwargs)
+
     @final
     async def async_run(self, *args: t.Any, **kwargs: t.Any) -> t.Any:
-        return await self._impl.async_run(*args, **kwargs)
+
+        self._execute_pre_run_hooks(*args, **kwargs)
```

Comment:
quick thoughts: It would be nice if we can execute pre_run hook in an async loop? 💭

Reply:
Indeed, though I think its irrelevant to discuss such implementation details until we understand how implementing this feature in the new runner's implementation architecture is possible